File: components/slider.ts
import { SliderModel } from "./icomponents.js";
import { Component } from "./component.js";
import { Change_t } from "./observable.js";


export class Slider extends Component<SliderModel> {
    render(): Element {
        let el = document.createElement("div");

        let slider = document.createElement("input");
        slider.type = "range";
        slider.min = this._model.min.toString();
        slider.max = this._model.max.toString();
        slider.value = this._model.val.toString();
        slider.step = this._model.step.toString();
        slider.addEventListener("input", () => this._model.val = Number(slider.value));

        let readout = document.createElement("input");
        readout.type = "number";
        readout.min = this._model.min.toString();
        readout.max = this._model.max.toString();
        readout.value = this._model.val.toString();
        readout.step = this._model.step.toString();
        readout.addEventListener("input", () => this._model.val = Number(readout.value));

        this._model.watch("val", (change: Change_t<number>) => readout.value = slider.value = change.new_value.toString());
        this._model.watch("min", (change: Change_t<number>) => slider.min = change.new_value.toString());
        this._model.watch("max", (change: Change_t<number>) => slider.max = change.new_value.toString());

        el.appendChild(slider);
        el.appendChild(readout);

        return el;
    }
}

File: app/slider.py
from typing import Any

from . import yg


class SliderModel(yg.Model):
    min: yg.Observable[float] = yg.obs(0.0)
    max: yg.Observable[float] = yg.obs(100.0)
    step: yg.Observable[float] = yg.obs(1.0)
    value: yg.Observable[float] = yg.obs(50.0)

    @yg.coerce(min)
    def _(self, __min: Any) -> float:
        return float(__min)

    @yg.validate(min)
    def _(self, __min: float) -> float:
        return min(__min, self.value.get())

    @yg.validate(max)
    def _(self, __max: float) -> float:
        return max(__max, self.value.get())

    @yg.validate(step)
    def _(self, __step: float) -> float:
        if __step <= 0:
            return self.step.get()
        return __step

    @yg.watch(step)
    def _step_change(self) -> None:
        value = self.value.get()
        step = self.step.get()
        self.value.set(round(value / step) * step)
